lib/Dialect/FIRRTL/Transforms/DropConst.cpp

```diff
@@ -0,0 +1,128 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the DropConst pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/FIRRTLUtils.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "llvm/Support/SaveAndRestore.h"
+
+#define DEBUG_TYPE "drop-const"
+
+using namespace circt;
+using namespace firrtl;
+
+// NOLINTBEGIN(misc-no-recursion)
+/// Return this type with a 'const' modifiers dropped
+static FIRRTLBaseType getAllConstDroppedType(FIRRTLBaseType type) {
+  if (!type.containsConst())
+    return type;
+  return TypeSwitch<FIRRTLBaseType, FIRRTLBaseType>(type)
+      .Case<ClockType, ResetType, AsyncResetType, AnalogType, SIntType,
+            UIntType>([](auto type) { return type.getConstType(false); })
+      .Case<BundleType>([](BundleType type) {
+        SmallVector<BundleType::BundleElement> constDroppedElements(
+            llvm::map_range(
+                type.getElements(), [](BundleType::BundleElement element) {
+                  element.type = getAllConstDroppedType(element.type);
+                  return element;
+                }));
+        return BundleType::get(type.getContext(), constDroppedElements, false);
+      })
+      .Case<FVectorType>([](FVectorType type) {
+        return FVectorType::get(getAllConstDroppedType(type.getElementType()),
+                                type.getNumElements(), false);
+      })
+      .Case<FEnumType>([](FEnumType type) {
+        SmallVector<FEnumType::EnumElement> constDroppedElements(
+            llvm::map_range(
+                type.getElements(), [](FEnumType::EnumElement element) {
+                  element.type = getAllConstDroppedType(element.type);
+                  return element;
+                }));
+        return FEnumType::get(type.getContext(), constDroppedElements, false);
+        ;
+      })
+      .Default([](Type) {
+        llvm_unreachable("unknown FIRRTL type");
+        return FIRRTLBaseType();
+      });
+}
+// NOLINTEND(misc-no-recursion)
+
+/// Returns null type if no conversion is needed.
+static FIRRTLBaseType convertType(FIRRTLBaseType type) {
+  auto nonConstType = getAllConstDroppedType(type);
+  return nonConstType != type ? nonConstType : FIRRTLBaseType{};
+}
+
+/// Returns null type if no conversion is needed.
+static Type convertType(Type type) {
+  if (auto base = type.dyn_cast<FIRRTLBaseType>()) {
+    return convertType(base);
+  }
+
+  if (auto refType = type.dyn_cast<RefType>()) {
+    if (auto converted = convertType(refType.getType()))
+      return RefType::get(converted, refType.getForceable());
+  }
```

Comment:
Can ref types go in aggregates? e.g., `{a : ref<uint<8>>}`

Reply:
They can only go into `Open` aggregates, which are lowered before this pass

Reply:
I should really look at where this gets inserted into the pipeline